rust/src/delta.rs

```diff
@@ -1355,6 +1455,223 @@ fn log_entry_from_actions(actions: &[Action]) -> Result<String, serde_json::Erro
     Ok(jsons.join("\n"))
 }
 
+/// Error returned when the CheckPointWriter is unable to write a checkpoint.
+#[derive(thiserror::Error, Debug)]
+pub enum CheckPointWriterError {
+    /// Error returned when the DeltaTableState does not contain a metadata action.
+    #[error("DeltaTableMetadata not present in DeltaTableState")]
+    MissingMetaData,
+    /// Error returned when creating the checkpoint schema.
+    #[error("DeltaLogSchemaError: {source}")]
+    DeltaLogSchema {
+        /// The source DeltaLogSchemaError
+        #[from]
+        source: DeltaLogSchemaError,
+    },
+    /// Passthrough error returned when calling DeltaTable.
+    #[error("DeltaTableError: {source}")]
+    DeltaTable {
+        /// The source DeltaTableError.
+        #[from]
+        source: DeltaTableError,
+    },
+    /// Error returned when the parquet writer fails while writing the checkpoint.
+    #[error("Failed to write parquet: {}", .source)]
+    ParquetError {
+        /// Parquet error details returned when writing the checkpoint failed.
+        #[from]
+        source: ParquetError,
+    },
+    /// Error returned when converting the schema to Arrow format failed.
+    #[error("Failed to convert into Arrow schema: {}", .source)]
+    ArrowError {
+        /// Arrow error details returned when converting the schema in Arrow format failed
+        #[from]
+        source: ArrowError,
+    },
+    /// Passthrough error returned when calling StorageBackend.
+    #[error("StorageError: {source}")]
+    Storage {
+        /// The source StorageError.
+        #[from]
+        source: StorageError,
+    },
+    /// Passthrough error returned by serde_json.
+    #[error("serde_json::Error: {source}")]
+    JSONSerialization {
+        /// The source serde_json::Error.
+        #[from]
+        source: serde_json::Error,
+    },
+}
+
+/// Struct for writing checkpoints to the delta log.
+pub struct CheckPointWriter {
+    table_uri: String,
+    delta_log_path: String,
+    last_checkpoint_path: String,
+    storage: Box<dyn StorageBackend>,
+    schema_factory: DeltaLogSchemaFactory,
+}
+
+impl CheckPointWriter {
+    /// Creates a new CheckPointWriter.
+    pub fn new(table_uri: &str, storage: Box<dyn StorageBackend>) -> Self {
+        let delta_log_path = storage.join_path(table_uri, "_delta_log");
+        let last_checkpoint_path = storage.join_path(delta_log_path.as_str(), "_last_checkpoint");
+        let schema_factory = DeltaLogSchemaFactory::new();
+
+        Self {
+            table_uri: table_uri.to_string(),
+            delta_log_path,
+            last_checkpoint_path,
+            storage,
+            schema_factory,
+        }
+    }
+
+    /// Creates a new checkpoint at the specified version.
+    /// NOTE: This method loads a new instance of delta table to determine the state to
+    /// checkpoint.
+    pub async fn create_checkpoint_for_version(
```

Comment:
I might be missing it, but is there any usage within delta-rs or users are expected to call it from within outside?

Reply:
For internal usage - see also https://github.com/delta-io/delta-rs/pull/280#issuecomment-856884265